### PR TITLE
Add documentation for Channel#send and Channel#close (crystal-lang#8346)

### LINKED ISSUE #8346: Behaviour of unbuffered Channel#send on Channel#close in v0.31.x vs previous versions
The Crystal compiler v0.31.x seems to encode a different behaviour than the previous versions in the scenario where
1. Fiber `f1` is waiting on a `#send` call on an unbuffered channel `c`
2. Fiber `f2` invokes `#close` on `c`.

### Outcome in v0.30.x and previous versions
`f1` gets unblocked as if another Fiber had received the value

### Outcome in v0.31.x
`#send` raises a `Channel::ClosedError` exception for `f1` to handle

This is exemplified in the code below.
```crystal
c = Channel(Int32).new
spawn do
  puts "f1 sending"
  c.send 2
  puts "f1 sent"
rescue
  puts "f1 rescued"
end
 
spawn do
  puts "f2 closing"
  c.close
  puts "f2 closed"
end
 
sleep 1
```
This prints the following in v0.30.0
```
#output 1
f1 sending
f2 closing
f2 closed
f1 sent
```
Whereas it prints the following in v0.31.x
```
#output 1
f1 sending
f2 closing
f2 closed
f1 rescued
```

Just wanted to figure out if this is intended or a regression, as this can have a pretty dramatic impact on error handling in concurrent apps. Happy to add some docs if we're happy with the current behaviour.

### PR DESCRIPTION
**Pull request by lbarasti**
Closes  #8346

### CODE REVIEW

File: src/channel.cr [hunk 1]
```diff
@@ -182,13 +191,15 @@ class Channel(T)
...
+  # If there is a value waiting, then it is returned immediately. Otherwise, this method blocks until a value is sent to the channel.
   #
   # Raises `ClosedError` if the channel is closed or closes while waiting for receive.
   #
   # ```
   # channel = Channel(Int32).new
-  # channel.send(1)
+  # spawn do
+  #   channel.send(1)
+  # end
```
Comment (lbarasti) [inline src/channel.cr]: An alternative way to fix the code snippet would be to go buffered
```crystal
channel = Channel(Int32).new(1)
channel.send(42)
channel.receive # => 42
```

File: src/channel.cr [hunk 2]
```diff
@@ -182,13 +191,15 @@ class Channel(T)
   end
 
   # Receives a value from the channel.
-  # If there is a value waiting, it is returned immediately. Otherwise, this method blocks until a value is sent to the channel.
+  # If there is a value waiting, then it is returned immediately. Otherwise, this method blocks until a value is sent to the channel.
```
Comment (lbarasti) [inline src/channel.cr]: Added `then` for consistency with the docs above - both are valid, I believe

File: src/channel.cr [hunk 3]
```diff
@@ -114,6 +114,10 @@ class Channel(T)
     end
   end
 
+  # Closes the channel.
+  # The method prevents any new value from being sent to / received from the channel.
+  # It wakes up any sender / receiver fibers waiting on the channel.
+  # It only has effect *after* any previously enqueued value in the channel has been received, up to the channel's capacity.
```
Comment (lbarasti) [inline src/channel.cr]: I appreciate the wording can be improved, so happy to take suggestions here

Comment (lbarasti) [inline src/channel.cr]: I would add that calling close on a closed channel does not have any effect, but such behaviour hasn't been documented in the specs, so I'd better add a test too, if we're keen on being explicit on that

File: src/channel.cr [hunk 4]
```diff
@@ -114,6 +114,10 @@ class Channel(T)
     end
   end
 
+  # Closes the channel.
+  # The method prevents any new value from being sent to / received from the channel.
+  # It wakes up any sender / receiver fibers waiting on the channel.
```
Comment (RX14) [inline src/channel.cr]: All fibers blocked in `send` or `receive` will recieve `Channel::ClosedError`?

File: src/channel.cr (hunk 3 above)
Comment (RX14) [reply src/channel.cr]: I'd probably just say

> All items successfully sent to the channel can be received before `receive` considers the channel closed, but `send` will consider the channel closed before `close` returns.

Comment (RX14) [reply src/channel.cr]: I'd document and spec that. You can do that in this PR.

File: src/channel.cr [hunk 5]
```diff
@@ -144,6 +148,11 @@ class Channel(T)
     @closed
   end
 
+  # Sends a value to the channel.
+  # If the channel has spare capacity, then the method returns immediately.
+  # Otherwise, this method blocks until the value is received from the channel or the channel's spare capacity becomes greater than 0.
```
Comment (RX14) [inline src/channel.cr]: > Otherwise, this method blocks the calling fiber until another fiber calls `receive`.

File: src/channel.cr (hunk 4 above)
Comment (lbarasti) [reply src/channel.cr]: Exactly, see
* [spec 1](https://github.com/crystal-lang/crystal/blob/master/spec/std/channel_spec.cr#L76) and [spec 2](https://github.com/crystal-lang/crystal/blob/master/spec/std/channel_spec.cr#L449) for `#receive`
* [spec](https://github.com/crystal-lang/crystal/blob/master/spec/std/channel_spec.cr#L143) for `#receive?`
* [spec 1](https://github.com/crystal-lang/crystal/blob/master/spec/std/channel_spec.cr#L217) and [spec 2](https://github.com/crystal-lang/crystal/blob/master/spec/std/channel_spec.cr#L427) for `#send`

File: src/channel.cr (hunk 3 above)
Comment (lbarasti) [reply src/channel.cr]: Done:

+ added to `Channel#close` docs
```
Calling `#close` on a closed channel does not have any effect.
```
+ Added spec
```
Channel "does not raise nor change its status when it is closed more than once"
```

File: src/channel.cr (hunk 4 above)
Comment (lbarasti) [reply src/channel.cr]: @RX14 
> All fibers blocked in send or receive will receive Channel::ClosedError

is ambiguous in that one might expect to actually `receive` an object of type ClosedError. I'm opting for

> All fibers blocked in send or receive will be awakened with Channel::ClosedError 

Let me know what you think

**Pull request merged**